### PR TITLE
optimize stack in the responder dispatch code path.

### PR DESCRIPTION
**Pull request by qizhangz**
Fix: #351 

Signed-off-by: Qi Zhang <qi1.zhang@intel.com>

### CODE REVIEW

File: include/internal/libspdm_common_lib.h [hunk 1]
```diff
@@ -285,7 +309,7 @@ typedef struct {
     // Cached plain text command
     // If the command is cipher text, decrypt then cache it.
     //
-    uint8_t last_spdm_request[MAX_SPDM_MESSAGE_BUFFER_SIZE];
+    void *last_spdm_request;
```
Comment (jyao1) [inline include/internal/libspdm_common_lib.h:312]: Please comment where it points to.

File: include/internal/libspdm_common_lib.h [hunk 2]
```diff
@@ -265,6 +265,30 @@ typedef struct {
...
+    // be SPDM transport message or secured SPDM transport message.
+    //
+    uint8_t request[MAX_SPDM_MESSAGE_BUFFER_SIZE];
+    //
+    // reserved bufeer for response in the main dispatch function
+    // in SPDM responder. This buffer is the transport message
+    // sent to spdm_context->send_message(). This message may
+    // be SPDM transport message or secured SPDM transport message.
+    //
+    uint8_t response[MAX_SPDM_MESSAGE_BUFFER_SIZE];
```
Comment (jyao1) [inline include/internal/libspdm_common_lib.h:281]: Please describe why we cannot merge those two buffer.

File: include/internal/libspdm_common_lib.h [hunk 3]
```diff
@@ -265,6 +265,30 @@ typedef struct {
...
+    //
+    // reserved buffer for secured_message in decode/encode
+    // function of transport layer message.
+    // This buffer is encrypted message in encode or decode function.
+    //
+    uint8_t secured_message[MAX_SPDM_MESSAGE_BUFFER_SIZE];
+
+    // This message is decrypted message in encode or decode function.
+    uint8_t app_message[MAX_SPDM_MESSAGE_BUFFER_SIZE];
```
Comment (jyao1) [inline include/internal/libspdm_common_lib.h:290]: Please follow the other comment to describe this data structure.

**Comment by jyao1** [comment]
Question: It seems only MCTP transport is updated. Do we also need change PCI_DOE ?

File: library/spdm_transport_mctp_lib/libspdm_mctp_mctp.c [hunk 4]
```diff
@@ -59,14 +62,28 @@ uint32_t spdm_mctp_get_max_random_number_count(void)
...
                   OUT void *transport_message)
 {
     uintn aligned_message_size;
     uintn alignment;
     mctp_message_header_t *mctp_message_header;
+    void *tmp_buffer;
+    uintn tmp_buffer_size;
+
+    // To requester message, use APP_message as tmp buffer to save message.
+    // because message and transport_message use the same buffer.
```
Comment (jyao1) [inline library/spdm_transport_mctp_lib/libspdm_mctp_mctp.c:79]: This is tricky, and documented in API header.
Or should the caller avoid this?

File: library/spdm_transport_mctp_lib/libspdm_mctp_mctp.c [hunk 5]
```diff
@@ -59,14 +62,28 @@ uint32_t spdm_mctp_get_max_random_number_count(void)
...
                   IN OUT uintn *transport_message_size,
                   OUT void *transport_message)
 {
     uintn aligned_message_size;
     uintn alignment;
     mctp_message_header_t *mctp_message_header;
+    void *tmp_buffer;
+    uintn tmp_buffer_size;
+
+    // To requester message, use APP_message as tmp buffer to save message.
```
Comment (jyao1) [inline library/spdm_transport_mctp_lib/libspdm_mctp_mctp.c:78]: "to requester message"
Do you mean a responder message?

**Review by jyao1 — APPROVED** [review]

File: include/internal/libspdm_common_lib.h [hunk 6]
```diff
@@ -265,6 +265,37 @@ typedef struct {
 #define spdm_context_struct_VERSION 0x1
 
 typedef struct {
+    //
+    // reserved buffer for request in the main dispatch function
+    // in SPDM responder. This buffer is the transport message
+    // recived from spdm_context->receive_message(). This message may
+    // be SPDM transport message or secured SPDM transport message.
+    // This buffer may be shared by last_spdm_request.
```
Comment (jyao1) [inline include/internal/libspdm_common_lib.h:273]: I think sharing with last_spdm_request may be problematic.

In SPDM 1.2, request message is transport layer message, while last_spdm_request is SPDM message, which might be larger than transport layer message if CHUNK is supported.

Can we copy it to app_message ?

File: library/spdm_transport_mctp_lib/libspdm_mctp_mctp.c [hunk 7]
```diff
@@ -149,8 +179,17 @@ return_status mctp_decode_message(OUT uint32_t **session_id,
         if (session_id != NULL) {
             *session_id = NULL;
         }
+        // To requester message, pass the SPDM message to last_spdm_request.
+        if (is_requester) {
+            libspdm_set_last_spdm_request(spdm_context, (uint8_t *)transport_message + sizeof(mctp_message_header_t),
```
[the remark this answers is not in the document]
Comment (rkongintel) [reply library/spdm_transport_mctp_lib/libspdm_mctp_mctp.c:184]: I agree this set of changes is overly complicated. 
Maybe should rethink about an alternate method.
Perhaps as a first pass, it is enough to move request/response buffers to context, and keep extra working temp buffers (secured/app buffers) on the stack.

**Comment by rkongintel** [comment]
I think this may be too complicated and error prone. 

Perhaps we should just start with moving just the request/response to context and leave other code the same?

**Comment by jyao1** [comment]
@rkongintel, I agree. I have same feeling.

For SPDM 1.2, the TransportMessageSize != MaxSPDMMessageSize. We need rework the flow anyway.
Maybe we can defer at that time.

**Comment by qizhangz** [comment]
So I simplified the change with with moving just the request/response to context. The stack is reduced by 8K size and context size is increased by 4K size.

**Comment by qizhangz** [comment]
@rkongintel Could you please review this PR again? Thank you!

**Comment by jyao1** [comment]
it seems OK as first step.
Let's keep #351 open and we need continue working on this.

**Review by steven-bellock — APPROVED** [review]

**Pull request merged**